stdlib/float.ml

```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
 let hash x = seeded_hash_param 10 100 0 x
 
 module Array = struct
+
   type t = floatarray
-  external create : int -> t = "caml_floatarray_create"
+
   external length : t -> int = "%floatarray_length"
   external get : t -> int -> float = "%floatarray_safe_get"
   external set : t -> int -> float -> unit = "%floatarray_safe_set"
+  external create : int -> t = "caml_floatarray_create"
   external unsafe_get : t -> int -> float = "%floatarray_unsafe_get"
   external unsafe_set : t -> int -> float -> unit = "%floatarray_unsafe_set"
+
+  let unsafe_fill a ofs len v =
+    for i = ofs to ofs + len - 1 do unsafe_set a i v done
+
+  let unsafe_blit src sofs dst dofs len =
+    for i = 0 to len - 1 do
+      unsafe_set dst (dofs + i) (unsafe_get src (sofs + i))
+    done
+
+  let check a ofs len msg =
+    if ofs < 0 || len < 0 || ofs + len < 0 || ofs + len > length a then
+      invalid_arg msg
+
+  let make n v =
+    let result = create n in
+    unsafe_fill result 0 n v;
+    result
+
+  let init l f =
+    if l < 0 then invalid_arg "Float.Array.init"
+    else
+      let res = create l in
+      for i = 0 to l - 1 do
+        unsafe_set res i (f i)
+      done;
+      res
+
+  let append a1 a2 =
+    let l1 = length a1 in
+    let l2 = length a2 in
+    let result = create (l1 + l2) in
+    unsafe_blit a1 0 result 0 l1;
+    unsafe_blit a2 0 result l1 l2;
+    result
+
+  (* next 3 functions: modified copy of code from string.ml *)
+  let ensure_ge (x:int) y =
+    if x >= y then x else invalid_arg "Float.Array.concat"
+
+  let rec sum_lengths acc = function
+    | [] -> acc
+    | hd :: tl -> sum_lengths (ensure_ge (length hd + acc) acc) tl
+
+  let concat l =
+    let len = sum_lengths 0 l in
+    let result = create len in
+    let rec loop l i =
+      match l with
+      | [] -> assert (i = len)
+      | hd :: tl ->
+        let hlen = length hd in
+        unsafe_blit hd 0 result i hlen;
+        loop tl (i + hlen)
+    in
+    loop l 0;
+    result
+
```

Comment:
Should we be worried performance-wise?
(`Array.`{`append`, `concat`} being basically
implemented in C, using `memcpy`)

Reply:
Is memcpy that much faster? I'll do some benchmarking.